presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java

```diff
@@ -0,0 +1,173 @@
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.prestosql.plugin.hive.HiveMetadata.PRESTO_QUERY_ID_NAME;
+import static io.prestosql.plugin.hive.HivePartitionManager.extractPartitionValues;
+import static io.prestosql.spi.StandardErrorCode.ALREADY_EXISTS;
+import static io.prestosql.spi.StandardErrorCode.INVALID_PROCEDURE_ARGUMENT;
+import static io.prestosql.spi.block.MethodHandleUtil.methodHandle;
+import static io.prestosql.spi.type.VarcharType.VARCHAR;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class RegisterPartitionProcedure
+        implements Provider<Procedure>
+{
+    private static final MethodHandle REGISTER_PARTITION = methodHandle(
+            RegisterPartitionProcedure.class,
+            "registerPartition",
+            ConnectorSession.class,
+            String.class,
+            String.class,
+            String.class,
+            List.class,
+            List.class);
+
+    private final Supplier<TransactionalMetadata> hiveMetadataFactory;
+    private final HdfsEnvironment hdfsEnvironment;
+    private final HiveMetastoreClosure metastore;
+
+    @Inject
+    public RegisterPartitionProcedure(Supplier<TransactionalMetadata> hiveMetadataFactory, HiveMetastore metastore, HdfsEnvironment hdfsEnvironment)
+    {
+        this.hiveMetadataFactory = requireNonNull(hiveMetadataFactory, "hiveMetadataFactory is null");
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.metastore = new HiveMetastoreClosure(requireNonNull(metastore, "metastore is null"));
+    }
+
+    @Override
+    public Procedure get()
+    {
+        return new Procedure(
+                "system",
+                "register_partition",
+                ImmutableList.of(
+                        new Procedure.Argument("schema_name", VARCHAR),
+                        new Procedure.Argument("table_name", VARCHAR),
+                        new Procedure.Argument("location", VARCHAR),
+                        new Procedure.Argument("partition_columns", new ArrayType(VARCHAR)),
+                        new Procedure.Argument("partition_values", new ArrayType(VARCHAR))),
+                REGISTER_PARTITION.bindTo(this));
+    }
+
+    public void registerPartition(ConnectorSession session, String schemaName, String tableName, String location, List<Object> partitionColumnNames, List<Object> partitionValues)
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
+            doRegisterPartition(session, schemaName, tableName, location, partitionColumnNames, partitionValues);
+        }
+    }
+
+    private void doRegisterPartition(ConnectorSession session, String schemaName, String tableName, String location, List<Object> partitionColumnNames, List<Object> partitionValues)
```

Comment:
What happens if any of arguments is NULL?

Reply:
It's not possible - it's validated when the procedure is called (https://github.com/prestosql/presto/blob/62ecc58e8667d58ac5907a9f8bc72540eb8eea18/presto-main/src/main/java/io/prestosql/execution/CallTask.java#L142).

That's why I've added https://github.com/prestosql/presto/pull/2706